File: nodejs/test/topjslibs/TopJsLibrariesTest.ts
import * as assert from "assert";
import {describe} from "mocha";
import {TopJsLibraries} from "../../src/topjslibs/TopJsLibraries";
import * as util from "./util";

describe("TopJsLibrariesTest", () => {
  describe("count(string)", () => {
    it("should return top 5 most popular JS libraries from the search result", () => {
      // Arrange
      const expected = new Array<[string, number]>(
          ["base.min;v35893530.js", 2],
          ["modules.min;v35893530.js", 2],
          ["webtk.min;v35893530.js", 2],
          ["webtrekk_cookieControl.min;v35893530.js", 2],
          ["application-0d7a7bcfe504533ad327.js", 1],
      );

      return new TopJsLibraries(5, 5, fetchHtml)
      // Act
      .count("test")
      // Assert
      .then((actual) => assert.deepStrictEqual(actual, expected));
    });
  });
});

function fetchHtml(url: string): Promise<string> {
  if (url === "https://www.google.com/search?q=test") {
    return new Promise<string>((resolve) => resolve(util.getHtmlFrom("google.com.html")));
  }

  if (url === "https://www.test.de/") {
    return new Promise<string>((resolve) => resolve(util.getHtmlFrom("test.de.html")));
  }
  if (url === "https://www.test.de/shop/test-hefte/") {
    return new Promise<string>((resolve) => resolve(util.getHtmlFrom("test-hefte.html")));
  }
  if (url === "https://www.test.de/thema/") {
    return new Promise<string>((resolve) => resolve(util.getHtmlFrom("thema.html")));
  }
  if (url === "https://www.oekotest.de/") {
    return new Promise<string>((resolve) => resolve(util.getHtmlFrom("oekotest.de.html")));
  }
  if (url === "https://de.wikipedia.org/wiki/Test_(Zeitschrift)") {
    return new Promise<string>((resolve) => resolve(util.getHtmlFrom("zeitschrift.html")));
  }

  return new Promise<string>((resolve) => resolve(""));
}

File: rust/src/top_js_libs/top_js_libraries_test.rs
use crate::top_js_libs::test_util::get_html_from;
use crate::top_js_libs::top_js_libraries::TopJsLibraries;
use futures::executor::block_on;

#[test]
fn count() {
    // Arrange
    let top_js_libraries = TopJsLibraries::new(5, 5, Box::from(fetch_html));
    let mut expected = Vec::new();
    expected.push(("base.min;v35893530.js".to_string(), 2));
    expected.push(("modules.min;v35893530.js".to_string(), 2));
    expected.push(("webtk.min;v35893530.js".to_string(), 2));
    expected.push(("webtrekk_cookieControl.min;v35893530.js".to_string(), 2));
    expected.push(("application-0d7a7bcfe504533ad327.js".to_string(), 1));

    // Act
    let actual = block_on(top_js_libraries.count("test"));

    // Assert
    assert_eq!(expected, actual);
}

fn fetch_html(url: &str) -> String {
    return match url {
        "https://www.google.com/search?q=test" => get_html_from("google.com.html"),

        "https://www.test.de/" => get_html_from("test.de.html"),
        "https://www.test.de/shop/test-hefte/" => get_html_from("test-hefte.html"),
        "https://www.test.de/thema/" => get_html_from("thema.html"),
        "https://www.oekotest.de/" => get_html_from("oekotest.de.html"),
        "https://de.wikipedia.org/wiki/Test_(Zeitschrift)" => get_html_from("zeitschrift.html"),

        _ => "".to_string()
    }
}